ringpop.thrift-gen

```diff
@@ -0,0 +1,171 @@
+{{ define "arguments" }}ctx {{ contextType }}{{ range .Arguments }}, {{ .Name }} {{ .Type | goType }}{{ end }}{{ end}}
+{{ define "call" }}ctx {{ range .Arguments }}, {{ .Name }}{{ end }}{{ end}}
+{{ define "returnType" }}{{ if .ReturnType }}r {{ .ReturnType | goType }}, {{ end }}err error{{ end}}
+
+{{ $Root := . }}
+package {{ .Package }}
+
+import (
+    "errors"
+    "fmt"
+    "sync"
+
+    "github.com/uber/ringpop-go"
+    "github.com/uber/tchannel-go"
+    "github.com/uber/tchannel-go/thrift"
+)
+
+{{ range .AST.Services }}
+{{ $Service := . }}
+{{ $TChanInterface :=  print "TChan" $Service.Name }}
+{{ $Adapter :=  print "Ringpop" $Service.Name "Adapter" }}
+{{ $ConfigName :=  print $Service.Name "Configuration" }}
+
+// ------ Generated code for {{ $Service.Name }} ------
+type {{ $Adapter }} struct {
+    impl        {{ $TChanInterface }}
+    ringpop     ringpop.Interface
+    ch          *tchannel.Channel
+    config      {{ $ConfigName }}
+    router      ringpop.Router
+}
+
+type {{ $ConfigName }} struct {
+    // structs for all service endpoints containing forwarding behavior
+    {{ range .Methods }}
+        {{ .Name | goPublicName }} *{{ $Service.Name }}{{ .Name | goPublicName }}Configuration
+    {{ end }}
+}
+// TODO make validation function on the config
+
+type tchan{{ $Service.Name }}Adaptable interface {
+    SetAdapter({{ $TChanInterface }})
+}
+
+func New{{ $Adapter }}(
+    impl   {{ $TChanInterface }},
+    rp     ringpop.Interface,
+    ch     *tchannel.Channel,
+    config {{ $ConfigName }},
+) *{{ $Adapter }} {
+    adapter := &{{ $Adapter }}{
+        impl:    impl,
+        ringpop: rp,
+        ch:      ch,
+        config:  config,
+    }
+    // create ringpop router for routing based on ring membership
+    adapter.router = ringpop.NewRouter(rp, adapter)
+
+    // notify implementation it is adapted
+    if adaptable, ok := impl.(tchan{{ $Service.Name }}Adaptable); ok {
+        adaptable.SetAdapter(adapter)
+    }
+
+    return adapter
+}
+
+// implement ringpop.ClientFactory
+func (a *{{ $Adapter }}) MakeClient(dest string) interface{} {
+    if dest == a.ringpop.WhoAmI() {
+        // use local client
+        return a.impl
+    }
+
+    // create remote client
+    client := thrift.NewClient(
```

Comment:
Could this logic live inside the router? Ideally, the only thing in the generated code would be the last line in this function, since nothing else is really specific to the adapter type